beforeInputPlugin: add native support for beforeinput

trueadm opened the pull request:
This PR adds support for the native `beforeinput` event (when we detected it), falling back to the alternatives if we don't find support for it. See https://github.com/facebook/react/issues/11211 for more context on this.

This PR should fix some long standing issues with DraftJS as well as providing better compatibility with modern browser events. This specific event is supported in all the major evergreen browsers, apart from Firefox.

TODO:
- Ensure we can use this internally without any issues
- Validate that the detection mechanism is reliable, or maybe switch to a better alternative

gaearon commented:
CI fails?

trueadm commented:
@gaearon Sorry I put this up quickly before going out for lunch. Hopefully tests should pass now.

johan commented:
Thanks again for working on this.

This is slight improvement over what currently exists, but it only wires up beforeinput events for some small subset of `inputType` actions the native `beforeinput` event caters to ([full list here](https://www.w3.org/TR/input-events-2/#interface-InputEvent-Attributes)).

Among the ones that work with the current branch are `insertText`, and some of the other insert by composition variants like if you type dead keys like ¨ followed by space or some vowel – but no deletion, cut/copy/paste or similar events seem to invoke the callback, at the moment.

If you want an interactive test rig to try your work on, and build your branch with `yarn build react/index,react-dom/index --type=UMD`, plug this into the `fixtures/packaging/babel-standalone/dev.html` script tag and open it in Safari or Chrome:

```js
      ReactDOM.render(
        <h1 contentEditable="true" suppressContentEditableWarning={true}
            onBeforeInput={(e) => {
          const ne = e.nativeEvent;
          const dt = ne.dataTransfer;
          const pasted = dt && dt.getData && dt.getData('text/html');
          let r = ne.getTargetRanges();
          r = r && r.map(({startOffset:a, endOffset:b}) => `${a}…${b}`).join();
          const info = `${e.type}:${ne.inputType} data:${e.data}, ranges:${r}`;
          console.info(info, pasted);
        }}>Click me to start editing!</h1>,
        document.getElementById('container')
      );
```

You should be able to delete, cut, copy, and paste, and see some rich inputType-annotated events for all of those actions, letting people build more serious editors from these primitives.

johan commented on packages/react-dom/src/events/plugins/__tests__/ModernBeforeInputEventPlugin-test.js:
```diff
@@ -825,4 +825,36 @@ describe('BeforeInputEventPlugin', () => {
+    }
+
+    it('should extract onBeforeInput on an input', () => {
+      const onBeforeInput = jest.fn();
```
Might be nice to actually test that the data gets passed through to the event handler; perhaps something like this:

```suggestion
      const onBeforeInput = jest.fn().mockImplementationOnce((e) => {
        expect(e.data).toBe('A');
      });
```

johan commented:
Beside the `type`, `data`, and `preventDefault` parts of the event already passed through to the react event, any advanced use of the interface will also want `inputType` and probably both the `getTargetRanges()` and `dataTransfer.getData(mimeType)` functions off of the native event, too.

As long as we have the nativeEvent property available, the current catering is enough to be usable, but I think updating `SyntheticEvent.js`'s `InputEventInterface` constant to something that doesn't reference the 2013 spec, but the union of `https://www.w3.org/TR/uievents/#event-type-beforeinput` and `https://www.w3.org/TR/input-events-2/#interface-InputEvent` instead would be excellent.

I'd at least add the `inputType` constant there, as it's effectively speaking the second half of the name of the set of events you subscribe to when asking for modern `beforeinput` and `input` events.

trueadm commented:
I've put this logic behind a flag for testing internally and I've addressed some feedback. @gaearon are you okay to move ahead with this? I feel it might address a Draft issue we are seeing internally with IME input.
